Sources/NIOSSL/SSLCertificateExtensions.swift

```diff
@@ -0,0 +1,221 @@
+        /// Otherwise `reference` could already be freed. This would result in undefined behaviour as we access a dangling pointer.
+        /// This method guarantees that `reference` is valid during execution of `body`.
+        internal func withReference<Result>(
+            _ body: (OpaquePointer?) throws -> Result
+        ) rethrows -> Result {
+            try withExtendedLifetime(self) {
+                switch self {
+                case .owned(let deallocator):
+                    return try body(deallocator.reference)
+                case .borrowed(let reference, _):
+                    return try body(reference)
+                }
+            }
+        }
+    }
+    
+    @usableFromInline internal let stackSize: Int
+    private let storage: Storage
+    
+    internal init(takeOwnershipOf reference: OpaquePointer?) {
+        self.storage = .init(takeOwnershipOf: reference)
+        if let reference = reference {
+            self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+        } else {
+            self.stackSize = 0
+        }
+    }
+    
+    internal init(borrowing reference: OpaquePointer?, owner: AnyObject) {
+        self.storage = .init(borrowing: reference, owner: owner)
+        if let reference = reference {
+            self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+        } else {
+            self.stackSize = 0
+        }
+    }
+    
+    internal init(copyOf reference: OpaquePointer?) {
+        guard let reference = reference else {
+            self.init()
+            return
+        }
+        self.init(takeOwnershipOf: CNIOBoringSSL_sk_X509_EXTENSION_dup(reference))
+    }
+    
+    internal init() {
+        self.init(takeOwnershipOf: nil)
+    }
+}
+
+extension NIOSSLCertificate {
+    public var _extensions: _NIOSSLCertificateExtensions {
+        _NIOSSLCertificateExtensions(borrowing: CNIOBoringSSL_X509_get0_extensions(self._ref), owner: self)
+    }
+}
+
+extension _NIOSSLCertificateExtensions: RandomAccessCollection {
+    public subscript(position: Int) -> _NIOSSLCertificateExtension {
+        precondition(self.indices.contains(position), "index \(position) out of bounds")
+        return self.storage.withReference { reference in
+            guard let reference = reference else {
+                fatalError("unexpected null pointer when unwrapping reference for value at \(position)")
+            }
+            guard let value = CNIOBoringSSLShims_sk_X509_EXTENSION_value(reference, position) else {
+                fatalError("unexpected null pointer when unwrapping extension value at \(position)")
+            }
+            
+            return .init(borrowing: value, owner: self.storage.owner)
+        }
+    }
+    
+    @inlinable public var startIndex: Int { 0 }
+    @inlinable public var endIndex: Int { self.stackSize }
+}
+
+public struct _NIOSSLCertificateExtension {
```

Comment:
I'm wondering if this and `_NIOSSLCertificateExtensions` should be nested types within `NIOSSLCertificate`.